### PR TITLE
overlord: add code to remove users that has an expiration date set (auto-removal 2/n)

### PR DESCRIPTION
**Pull request by Meulengracht**
Second PR in the auto-removal of users PR series. This PR builds upon https://github.com/snapcore/snapd/pull/12079 and adds the overlord specific bits, which implements the removal logic as a part of Ensure() which is called periodically. We also add the 'Expiration' member to the UserState.

### CODE REVIEW

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1438]: nitpick: I'd suggest putting this near the top of the file

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1440]: Could this be called `removeUser` instead, since it does not use the fact that the user has expired?

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1447]: I would expect these two function calls to be swapped. If I'm reading the code right, in `daemon/api_users.go` we first call `osutilAddUser()` and later call `setupLocalUser()`, so it seems more logical if the delition happens in the reverse order. If there are reasons why it's shouldn't be done in that way, it would be nice to add a code comment with them.

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1428]: nitpick: `userHasExpired` (or "is expired?" - not sure!)

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1453]: just `removeExpiredUsers`?

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1458]: As far as I can tell from other parts of the code, it seems to me that "on classic" only changes the location of the users' database (on core, we need to use the `--extrausers` flag), but the fact that users are added remains valid on classic too. But better wait for Samuele.

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go:1447]: In `daemon/api_users.go` the deletion occurs in this order, but I can see your point. There is no explanation as to why the user is deleted in this order. My theory is that there might be some conditions under where the user cannot be removed/fails to be removed from the system, and if the user was *not* removed by the system, then it's crucial it not be removed from auth state.

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Thank you for the suggestion, updated this!

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Good point!

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Updated, thank you!

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go:1455]: I wanted to keep the naming convention as it was a part of ensure, but I'm not sure if that's important or if I'm just misunderstanding how we use the term, i'll happily change it if I'm using it incorrectly here

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go:1460]: I agree! @pedronis

File: overlord/devicestate/devicemgr.go
Comment (mardy) [reply overlord/devicestate/devicemgr.go]: OK, fine with me then. But let's leave this conversation open so that Samuele can see it and weigh in if needed.

File: overlord/devicestate/devicemgr.go
Comment (mardy) [reply overlord/devicestate/devicemgr.go:1455]: Let's leave this to Samuele then. I suggested my preference, but it's not important :-)

**Comment by pedronis** [comment]
@Meulengracht  we have this open as well:

https://github.com/snapcore/snapd/pull/11796

I think we want to land that one and then the code here can be reorganized a bit

File: overlord/devicestate/devicemgr.go
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1451]: these don't use state on DeviceManager so a bit unclear why make them methods here, also in the PR I mentioneds we grow a new users.go file which is probably the right place for some of this.

File: overlord/devicestate/devicestate_systems_test.go
Comment (pedronis) [inline overlord/devicestate/devicestate_systems_test.go:1301]: this is not the right test file for this, as it's not about systems functionality which relates to seed systems

**Review by pedronis — COMMENTED** [review]
did a quick pass, but as I wrote we probably want work on landing the older PR first

File: overlord/auth/auth.go
Comment (pedronis) [inline overlord/auth/auth.go:145]: already landed but I would prefer this to be called NewUserParams

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Resolving this issue as this code has been removed

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: I removed the check as you said the users added remain valid on classic, I instead refactored the test to verify that `--extrausers` flag is correctly set/not set on core/classic.

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Updated this after Ondrej's PR was merged!

File: overlord/devicestate/devicestate_systems_test.go
Comment (Meulengracht) [reply overlord/devicestate/devicestate_systems_test.go]: I moved them to devicestate_test.go instead and changed the tests a bit.

File: overlord/auth/auth.go
Comment (Meulengracht) [reply overlord/auth/auth.go:145]: Thanks for the feedback here! I added this to my TODO as a followup. It introduces to many changed files to do in this PR.

File: overlord/devicestate/devicemgr.go
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1563]: This comment and the one below are questions, yet describe exactly what the code is doing. If you want to draw the reviewer's attention, I would write a comment here in github.

At this point I think you can remove the two comments; if you decide to keep them anyways, I'd add a bit of more information in them, to explain why we might want to do otherwise.

**Review by mardy — APPROVED** [review]
LGTM, just a nitpick about unclear comments.

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Good point, updated the comments. Thanks!

File: overlord/auth/auth.go
Comment (Meulengracht) [reply overlord/auth/auth.go:145]: https://github.com/snapcore/snapd/pull/12226

File: overlord/auth/auth.go
Comment (pedronis) [inline overlord/auth/auth.go:187]: it seems we need a new NewUser test to cover this bit?

File: overlord/devicestate/devicemgr.go
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1546]: this isn't using anything from DeviceManager, it seems more an helper that could live in users.go or even become a method on UserState itself?

File: overlord/devicestate/devicemgr.go
Comment (pedronis) [reply overlord/devicestate/devicemgr.go:1558]: no, it's correct, all methods contributing to Ensure behavior and called from it start with ensure...

File: overlord/devicestate/devicemgr.go
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1565]: this should use SysAny if the intention is really to work the same on also classic systems

File: overlord/devicestate/devicestate_test.go
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go:2289]: the classic test shouldn't set mode

File: overlord/devicestate/devicestate_test.go
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go:2322]: seems this test and the previous could share code via a testEnsure... test method?

File: overlord/devicestate/devicestate_test.go
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go:2407]: same for these other tests checking that DelUser is **not** called

**Review by pedronis — CHANGES_REQUESTED** [review]
thanks, something is off with the mode check if the intention is really for this to work anywhere (classic, core). Some other comments

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go:1546]: I like the suggestion about integrating it into UserState itself, i'll do that!

File: overlord/auth/auth.go
Comment (Meulengracht) [reply overlord/auth/auth.go:187]: Yea very nice catch, definitely missed that one

File: overlord/devicestate/devicestate_test.go
Comment (Meulengracht) [reply overlord/devicestate/devicestate_test.go]: Yea that makes a lot of sense, I fixed this, thanks for the suggestion!

File: overlord/devicestate/devicestate_test.go
Comment (Meulengracht) [reply overlord/devicestate/devicestate_test.go]: Also done, thanks!

File: overlord/devicestate/devicestate_test.go
Comment (Meulengracht) [reply overlord/devicestate/devicestate_test.go]: Ah right, I should have tested if it worked without! Thank you!

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Thanks for this, I changed this but I don't think I understand the Modeenv correctly - Sent you a DM.

File: overlord/devicestate/devicemgr.go
Comment (Meulengracht) [reply overlord/devicestate/devicemgr.go]: Fixed

File: overlord/auth/auth.go
Comment (mardy) [inline overlord/auth/auth.go:88]:
```suggestion
// the expiration date is earlier than the current time.
```
("past" as a preposition usually means "after", so this is ambiguous)

**Review by mardy — APPROVED** [review]
LGTM, just a nitpick :-)

File: overlord/auth/auth.go
Comment (pedronis) [reply overlord/auth/auth.go:88]: I think it's probably clearer switched around:

"... and we are paste the expiration date"
"... and current time is past the expiration date"

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/devicestate/devicestate_test.go
Comment (Meulengracht) [reply overlord/devicestate/devicestate_test.go]: Thanks! Changed this, and good to know

**Pull request merged**
